### PR TITLE
Functional gas estimator

### PR DESCRIPTION
**Pull request by chriseth**
This adds the feature to estimate the gas costs of a specific function of a contract on arbitrary input.

### CODE REVIEW

File: mix/CodeModel.cpp [hunk 1]
```diff
@@ -348,8 +348,7 @@ void CodeModel::gasEstimation(solidity::CompilerStack const& _cs)
 			continue;
 		dev::solidity::SourceUnit const& sourceUnit = _cs.getAST(*contractDefinition.getLocation().sourceName);
 		AssemblyItems const* items = _cs.getRuntimeAssemblyItems(n);
-		StructuralGasEstimator estimator;
-		std::map<ASTNode const*, GasMeter::GasConsumption> gasCosts = estimator.breakToStatementLevel(estimator.performEstimation(*items, std::vector<ASTNode const*>({&sourceUnit})), {&sourceUnit});
+		std::map<ASTNode const*, GasMeter::GasConsumption> gasCosts = GasEstimator::breakToStatementLevel(GasEstimator::structuralEstimation(*items, std::vector<ASTNode const*>({&sourceUnit})), {&sourceUnit});
```
Comment (LianaHus) [inline mix/CodeModel.cpp]: I think there is no need for "std::" here

File: libevmasm/GasMeter.cpp [hunk 2]
```diff
@@ -29,12 +29,13 @@ using namespace dev::eth;
 
 GasMeter::GasConsumption& GasMeter::GasConsumption::operator+=(GasConsumption const& _other)
 {
-	isInfinite = isInfinite || _other.isInfinite;
+	if (_other.isInfinite && !isInfinite)
+		*this = infinite();
 	if (isInfinite)
 		return *this;
 	bigint v = bigint(value) + _other.value;
 	if (v > std::numeric_limits<u256>::max())
```
Comment (LianaHus) [inline libevmasm/GasMeter.cpp]: std::

File: test/libsolidity/GasMeter.cpp [hunk 3]
```diff
@@ -47,30 +48,47 @@ class GasMeterTestFramework: public ExecutionFramework
...
+		GasMeter::GasConsumption gas = meter.estimateMax(0, state);
+		u256 bytecodeSize(m_compiler.getRuntimeBytecode().size());
 		gas += bytecodeSize * c_createDataGas;
 		BOOST_REQUIRE(!gas.isInfinite);
 		BOOST_CHECK(gas.value == m_gasUsed);
 	}
 
+	/// Compares the gas computed by PathGasMeter for the given signature (but unknown arguments)
+	/// against the actual gas usage computed by the VM on the given set of argument variants.
+	void testRunTimeGas(std::string const& _sig, vector<bytes> _argumentVariants)
```
Comment (LianaHus) [inline test/libsolidity/GasMeter.cpp]: std::

File: mix/CodeModel.cpp (hunk 1 above)
Comment (chriseth) [inline mix/CodeModel.cpp]: There is no "using std" probably because of the many name clashes between Qt and std.

**Pull request merged**